### PR TITLE
8315024: Vector API FP reduction tests should not test for exact equality

### PR DESCRIPTION
**Pull request by gergo-**
Certain floating point reduction operations in the Vector API are allowed to introduce rounding errors. Adjust the corresponding tests to allow a small relative error when comparing the operation's result to the expected value. Also, add a new generator `double[i / 10.0 + 0.1]` to test floating point operations with somewhat more interesting input data.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315024](https://bugs.openjdk.org/browse/JDK-8315024): Vector API FP reduction tests should not test for exact equality (**Bug** - P5)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16024/head:pull/16024` \
`$ git checkout pull/16024`

Update a local copy of the PR: \
`$ git checkout pull/16024` \
`$ git pull https://git.openjdk.org/jdk.git pull/16024/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16024`

View PR using the GUI difftool: \
`$ git pr show -t 16024`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16024.diff">https://git.openjdk.org/jdk/pull/16024.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16024#issuecomment-1744471104)

### CODE REVIEW

File: test/jdk/jdk/incubator/vector/Double128VectorTests.java
Comment (eme64) [inline test/jdk/jdk/incubator/vector/Double128VectorTests.java:64]: It seems a bit large. Have you tried to make it smaller? Or what is your justification for this value?

File: test/jdk/jdk/incubator/vector/Double128VectorTests.java
Comment (eme64) [inline test/jdk/jdk/incubator/vector/Double128VectorTests.java:150]: Optional: reduce code duplication. Call `assertReductionArraysEquals` with `relativeError = 0` in pre-existing method.

File: test/jdk/jdk/incubator/vector/Double128VectorTests.java
Comment (eme64) [inline test/jdk/jdk/incubator/vector/Double128VectorTests.java:1117]: Did this generate rounding issues for addition?
Another option would be random values that make sure to fill the whole mantissa with random information.
Of course the tricky part is to keep them within reasonable bounds so that on multiplication they do not degenerate to zero or infinity.
Also: it would be nice to have some cases with extreme values (infinity, NaN, etc).

**Review by eme64 — CHANGES_REQUESTED** [review]
Generally looks good, thanks for looking into this.
I left a few comments below.

Another concern I have, which I ran into by writing tests for the auto-vectorizer:
Are we making sure the float/double reductions do not degenerate to either zero or infinity? Because if they do degenerate, then we have only a very weak test.
I'm especially worried about all the values that depend on `i`, and then get multiplied. Don't the multiplications hit the maximal float value very quickly?

**Comment by gergo-** [comment]
> Are we making sure the float/double reductions do not degenerate to either zero or infinity?

You're right. We're currently getting "lucky" here because we reduce individual vectors of 4, 8, or 16 elements inside a loop. So even if the array contains a zero, or the product of the entire array is infinite, individual blocks inside it will have finite, non-zero products. This will change when https://bugs.openjdk.org/browse/JDK-8309647 is addressed. I'll try to put together a generator that also works nicely for multiplications over a whole array.

File: test/jdk/jdk/incubator/vector/Double128VectorTests.java
Comment (gergo-) [reply test/jdk/jdk/incubator/vector/Double128VectorTests.java:1117]: Yes, this generator generates a rounding issue for addition:
```
test FloatMaxVectorTests.ADDReduceFloatMaxVectorTests(float[i / 10.0 + 0.1]): failure
java.lang.AssertionError: at index #16 expected [39.2] but found [39.199997]
```

File: test/jdk/jdk/incubator/vector/Double128VectorTests.java
Comment (eme64) [reply test/jdk/jdk/incubator/vector/Double128VectorTests.java:64]: I'm just worried that 1% is very much for the addition tests. Basically we might be dropping off a whole element, and would not notice it.

File: test/jdk/jdk/incubator/vector/Double128VectorTests.java
Comment (gergo-) [reply test/jdk/jdk/incubator/vector/Double128VectorTests.java:1103]: I updated this generator to `0.01 + (i / (i + 1))`, plus a variant that replaces some of the elements of this sequence with values from the `cornerCaseValues` generator.
This should address all of your concerns. Almost all values in this sequence are very close to 1, so they can be added and multiplied without overflow, up to about 2000 elements for `float`s. The mantissas have bits all over the place. The exponents are very limited, but a wider range of exponents is exercised by the other tests.

**Comment by gergo-** [comment]
@eme64 would you have time to take another look at the changes I have made to this PR?

**Comment by eme64** [comment]
@gergo- I just looked at it again. It looks better.

Still, I have a concern about `cornerCaseValues`: Does the result not always degenerate to zero / NaN now?
Every 17th value is a corner-case. And there, we mix in all of the special cases deterministically (zero, infty, min/max, NaN), depending on INT/FP type. But it seems that way we always have a NaN in the FP array - and therefore the reduction would always be NaN. Similar things happen with int-multiplication and zero.

I wonder if it would not be better to generate some data randomly, and throw in the special-cases with a very low probability. Maybe 50% that any show up, and then randomly pick one or more special case values. That way you can test the different special-cases separately. And their position could also be random.

BTW: is there any wiki about the template file format, and how to "compile" it to java? I might want to use it in the future myself :)

**Comment by gergo-** [comment]
> @gergo- I just looked at it again. It looks better.

Thanks.

> Still, I have a concern about `cornerCaseValues`: Does the result not always degenerate to zero / NaN now? Every 17th value is a corner-case. And there, we mix in all of the special cases deterministically (zero, infty, min/max, NaN), depending on INT/FP type. But it seems that way we always have a NaN in the FP array - and therefore the reduction would always be NaN. Similar things happen with int-multiplication and zero.

Currently there are reduction tests that reduce not across the whole input array but over individual vector-sized blocks, e.g.:
```
        for (int ic = 0; ic < INVOC_COUNT; ic++) {
            for (int i = 0; i < a.length; i += SPECIES.length()) {
                DoubleVector av = DoubleVector.fromArray(SPECIES, a, i);
                r[i] = av.reduceLanes(VectorOperators.MUL);
            }
        }
```

The largest vector length is 16 elements (32 bit float * 16 = 512 bits max vector size). Therefore at most every second block will contain one corner-case value, and all the other blocks will only contain normal values. No block mixes different corner case values.

> I wonder if it would not be better to generate some data randomly, and throw in the special-cases with a very low probability. Maybe 50% that any show up, and then randomly pick one or more special case values. That way you can test the different special-cases separately. And their position could also be random.

I would say that this could be tackled more naturally as part of https://bugs.openjdk.org/browse/JDK-8309647 which concerns moving reductions out of loops and would require revisiting these tests anyway.

> BTW: is there any wiki about the template file format, and how to "compile" it to java? I might want to use it in the future myself :)

I'm not aware of any docs, I learned to do this by doing. I just do `bash gen-tests.sh` to generate the Java code.

**Review by eme64 — APPROVED** [review]
@gergo- Sounds ok. We don't have to do all in your change here. Thanks for the work you did!

**Comment by gergo-** [comment]
Thanks for the review @eme64 !

/integrate

**Comment by eme64** [comment]
@gergo- This PR will require a second review before being sponsored. It is not trivial.

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me.

**Pull request closed without merging**
